crypto/fipsmodule/aes/asm/aesni-avx512.pl

```diff
@@ -0,0 +1,3165 @@
+      mov     $TWTEMPH, 0x48($TW)
+___
+    }
+
+    # round 7
+    $code .= "vmovdqa 0xf0($TW), $t0\n";
+
+    for (my $i = 0; $i < $num_blocks; $i++) {
+      $code .= "vaesdec $t0, $st[$i], $st[$i]\n";
+    }
+
+    if (0 == $lt128) {
+      $code .= <<___;
+      xor     $ghash_poly_8b_temp, $ghash_poly_8b_temp
+      shl     \$1, $TWTEMPL
+      adc     $TWTEMPH, $TWTEMPH
+      cmovc   $ghash_poly_8b, $ghash_poly_8b_temp
+      xor     $ghash_poly_8b_temp, $TWTEMPL
+      mov     $TWTEMPL, 0x50($TW) # next Tweak6 generated
+      mov     $TWTEMPH, 0x58($TW)
+___
+    }
+
+    # round 8
+    $code .= "vmovdqa 0x100($TW), $t0\n";
+
+    for (my $i = 0; $i < $num_blocks; $i++) {
+      $code .= "vaesdec $t0, $st[$i], $st[$i]\n";
+    }
+
+    if (0 == $lt128) {
+      $code .= <<___;
+      xor     $ghash_poly_8b_temp, $ghash_poly_8b_temp
+      shl     \$1, $TWTEMPL
+      adc     $TWTEMPH, $TWTEMPH
+      cmovc   $ghash_poly_8b, $ghash_poly_8b_temp
+      xor     $ghash_poly_8b_temp, $TWTEMPL
+      mov     $TWTEMPL, 0x60($TW) # next Tweak7 generated
+      mov     $TWTEMPH, 0x68($TW)
+___
+    }
+
+    # round 9
+    $code .= "vmovdqa 0x110($TW), $t0\n";
+
+    for (my $i = 0; $i < $num_blocks; $i++) {
+      $code .= "vaesdec $t0, $st[$i], $st[$i]\n";
+    }
+
+    if (0 == $lt128) {
+      $code .= <<___;
+      xor     $ghash_poly_8b_temp, $ghash_poly_8b_temp
+      shl     \$1, $TWTEMPL
+      adc     $TWTEMPH, $TWTEMPH
+      cmovc   $ghash_poly_8b, $ghash_poly_8b_temp
+      xor     $ghash_poly_8b_temp, $TWTEMPL
+      mov     $TWTEMPL, 0x70($TW) # next Tweak8 generated
+      mov     $TWTEMPH, 0x78($TW)
+___
+    }
+
+    # round 10
+    $code .= "vmovdqa 0x120($TW), $t0\n";
+
+    for (my $i = 0; $i < $num_blocks; $i++) {
+      $code .= "vaesdec $t0, $st[$i], $st[$i]\n";
+    }
+
+    # round 11
+    $code .= "vmovdqa 0x130($TW), $t0\n";
+
+    for (my $i = 0; $i < $num_blocks; $i++) {
+      $code .= "vaesdec $t0, $st[$i], $st[$i]\n";
+    }
+
+    # round 12
+    $code .= "vmovdqa 0x140($TW), $t0\n";
+
+    for (my $i = 0; $i < $num_blocks; $i++) {
+      $code .= "vaesdec $t0, $st[$i], $st[$i]\n";
+    }
+
+    # round 13
+    $code .= "vmovdqa 0x150($TW), $t0\n";
+
+    for (my $i = 0; $i < $num_blocks; $i++) {
+      $code .= "vaesdec $t0, $st[$i], $st[$i]\n";
+    }
+
+    # round 14
+    $code .= "vmovdqa 0x160($TW), $t0\n";
+
+    for (my $i = 0; $i < $num_blocks; $i++) {
+      $code .= "vaesdeclast $t0, $st[$i], $st[$i]\n";
+    }
+
+    # xor Tweak values
+    for (my $i = 0; $i < $num_blocks; $i++) {
+      $code .= "vpxor $tw[$i], $st[$i], $st[$i]\n";
+    }
+
+    if (0 == $lt128) {
+      # load next Tweak values
+      $code .= <<___;
+      vmovdqa  ($TW), $tw1
+      vmovdqa  0x10($TW), $tw2
+      vmovdqa  0x20($TW), $tw3
+      vmovdqa  0x30($TW), $tw4
+      vmovdqa  0x40($TW), $tw5
+      vmovdqa  0x50($TW), $tw6
+      vmovdqa  0x60($TW), $tw7
+___
+    }
+  }
+
+  # Encrypt 8 blocks in parallel
```

Comment:
The number of blocks encrypted (decrypted) in parallel in a loop iteration is 8 or 16?

Reply:
it has both, I will update the comment on encrypt_by_16_zmm.